lib/nostrum/struct/user.ex

```diff
@@ -42,27 +56,18 @@ defmodule Nostrum.Struct.User do
     email: email,
   }
 
-  @derive [Poison.Encoder]
-  defstruct [
-    :id,
-    :username,
-    :discriminator,
-    :avatar,
-    :bot,
-    :mfa_enabled,
-    :verified,
-    :email,
-  ]
-
   @doc false
   def p_encode do
     %__MODULE__{}
   end
 
   @doc false
-  # TODO: Necessary?
-  def to_struct(%{__struct__: _} = map), do: map
   def to_struct(map) do
-    struct(__MODULE__, map)
+    new = 
+      map 
+      |> Map.new(fn {k, v} -> {Util.maybe_to_atom(k), v} end) 
```

Comment:
I'm probably missing something but what's the purpose of this?

Reply:
Line 68 converts all of the map's keys into atoms using `maybe_to_atom/1` from the Util module.

There is a function in Util that already does something like this (`safe_atom_map/1` off the top of my head), but I did not want to use it due to having a weird side effect. That function attempts to convert all map keys into atoms, including inner maps as well. I think `safe_atom_map/1` is a problem for `to_struct/1` functions, since we don't want to convert all of the inner map keys into atoms. For example, using `safe_atom_map/1` on a `Nostrum.Struct.Guild` struct will affect each of the "indexed" data fields in the struct. Off the top of my head, these are `:members`, `:channels`, and `:roles`. For these fields, `safe_atom_map/1` will attempt to convert their ids (4317843914731) into atoms.

Rather than change `safe_atom_map/1`, I'd just do what I did in line 68 for a "flat" atom-map function.

Reply:
Thanks for the explanation, no problem with this then.